File: scripts/run_experience_trace_tests.mjs
#!/usr/bin/env node
import assert from "node:assert/strict";
import crypto from "node:crypto";
import fs from "node:fs/promises";
import os from "node:os";
import path from "node:path";
import { beginExperienceTrace } from "./lib/experience_trace.mjs";

function canonical(value) {
  if (Array.isArray(value)) return value.map(canonical);
  if (value && typeof value === "object") {
    return Object.fromEntries(Object.keys(value).sort().map((key) => [key, canonical(value[key])]));
  }
  return value;
}
function sha(value) {
  return crypto.createHash("sha256").update(JSON.stringify(canonical(value))).digest("hex");
}

const root = await fs.mkdtemp(path.join(os.tmpdir(), "excel-inflow-trace-test-"));
const output = path.join(root, "trace.json");
delete process.env.EXCEL_INFLOW_EXPERIENCE_TRACE_ID;
delete process.env.EXCEL_INFLOW_EXPERIENCE_TRACE_PATH;
const trace = beginExperienceTrace({ runId: "trace_test", outputPath: output });
const broker = trace.startSpan({ component: "broker", phase: "native_inspection", attributes: { houses: 3 } });
trace.endSpan(broker, { outcome: "PASS", attributes: { native_clean_houses: 1 } });
const host = trace.startSpan({ component: "model_host", phase: "semantic_response", waitReason: "external_host" });
trace.endSpan(host, { outcome: "DEGRADED" });
trace.markVisibleResponse({ status: "COMPLETE", attributes: { workbook_published: true } });
trace.finalize({ outcome: "PASS" });
const value = JSON.parse(await fs.readFile(output, "utf8"));
assert.equal(value.schema_version, "excel-inflow-experience-trace/1.0");
assert.equal(value.trace_id, trace.traceId);
assert.equal(value.run_id, "trace_test");
assert.equal(value.status, "PASS");
assert.ok(value.visible_response_at);
assert.equal(value.spans.length, 3);
assert.equal(value.spans[0].phase, "user_submission_to_visible_response");
assert.ok(value.spans.slice(1).every((span) => span.parent_span_id === value.root_span_id));
assert.ok(value.spans.every((span) => span.completed_at));
const unsigned = { ...value, trace_sha256: null };
assert.equal(value.trace_sha256, sha(unsigned));
const carrierSource = await fs.readFile(new URL("./lib/run_carrier.mjs", import.meta.url), "utf8");
assert.ok(carrierSource.includes("experience_trace_id"));
const controllerSource = await fs.readFile(new URL("./run_excel_inflow_vnext.mjs", import.meta.url), "utf8");
assert.ok(controllerSource.includes("experience_trace"));
console.log(JSON.stringify({ status: "PASS", checks: 14, trace: output }, null, 2));

File: scripts/run_experience_trace_tests.py
#!/usr/bin/env python3
import time
from experience_trace import ExperienceTrace, WARNING_MS, INVESTIGATION_MS, BLOCK_MS, coverage_summary
assert (WARNING_MS,INVESTIGATION_MS,BLOCK_MS)==(30000,120000,300000)
t=ExperienceTrace('trace-test')
with t.span('owned','test'): time.sleep(.005)
r=t.finish(); assert r['schema_version']=='experience-trace/1.0'; assert r['summary']['classification_ratio']>0.7; assert r['summary']['certification_block_gap_count']==0
synthetic=coverage_summary([
    {'span_id':'root','parent_span_id':None,'category':'excel_inflow_active','external_wait_reason':None,'start_offset_ms':0,'end_offset_ms':100,'metadata':{'coverage_role':'root'}},
    {'span_id':'stage','parent_span_id':'root','category':'excel_inflow_active','external_wait_reason':None,'start_offset_ms':10,'end_offset_ms':90,'metadata':{}},
    {'span_id':'leaf','parent_span_id':'stage','category':'excel_inflow_active','external_wait_reason':None,'start_offset_ms':20,'end_offset_ms':40,'metadata':{}},
    {'span_id':'wait','parent_span_id':'stage','category':'known_external_wait','external_wait_reason':'host','start_offset_ms':60,'end_offset_ms':80,'metadata':{}},
],100)
assert synthetic['leaf_span_ids']==['leaf','wait']
assert synthetic['classified_ms']==40 and synthetic['unknown_ms']==60
assert coverage_summary([{'span_id':'root','parent_span_id':None,'category':'controller_root','external_wait_reason':None,'start_offset_ms':0,'end_offset_ms':100,'metadata':{}}],100)['classified_ms']==0
print('{"status":"PASS","checks":8,"mutations_rejected":1}')
